### PR TITLE
[wxwidgets] fix wxWidgets_LIBRARIES for windows, incl. static builds

### PR DESCRIPTION
**Pull request by jjYBdx4IL**
- #### What does your PR fix?
 * fix wxWidgets_LIBRARIES for windows static builds (add missing comctl32 etc.)
 * add wxbase to list of debug libs

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
only changed windows stuff, except for the added wxbase debug lib. Is there any convenient CI system available to test the other platforms/triplets before creating a pull request?

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes. Think so.

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**
No.

### CODE REVIEW

**Comment by cenit** [comment]
appending to a file provided by vcpkg itself is not so nice. isn’t it better to make it a configurable file and copy it through a configure_file “@only” with appropriate variables?

**Comment by jjYBdx4IL** [comment]
> appending to a file provided by vcpkg itself is not so nice. isn’t it better to make it a configurable file and copy it through a configure_file “@only” with appropriate variables?

Indeed. I wasn't aware of that mechanism, yet.

**Review by LilyWangLL — APPROVED** [review]

File: ports/wxwidgets/portfile.cmake [hunk 1]
```diff
@@ -128,7 +128,26 @@ if(NOT EXISTS "${CURRENT_PACKAGES_DIR}/include/wx/setup.h")
     configure_file("${CMAKE_CURRENT_LIST_DIR}/setup.h.in" "${CURRENT_PACKAGES_DIR}/include/wx/setup.h" @ONLY)
 endif()
 
-file(COPY "${CMAKE_CURRENT_LIST_DIR}/vcpkg-cmake-wrapper.cmake" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+# insert a fixed set of properly categorized libs into the cmake config
+file(GLOB WX_DEBUG_LIBS "${CURRENT_PACKAGES_DIR}/debug/lib/wx*31*")
+file(GLOB WX_OPT_LIBS "${CURRENT_PACKAGES_DIR}/lib/wx*31*")
```
Comment (dg0yt) [inline ports/wxwidgets/portfile.cmake:133]: To be more consistent with naming in vcpkg, I would prefer `WX_LIBS_DEBUG`/`WX_LIBS_RELEASE`.

**Review by dg0yt — COMMENTED** [review]
Basing the link libraries on globbing will break most static non-MSVC cases because the order of libraries matters for these linkers.

Unfortunately, we lack some CI coverage here, due to WxWidgets' dependency on gtk3 for Linux. There must be manual testing on x64-linux with some port that uses this wrapper.
(wxwidgets:x64-linux was fixed only recently, and a immediate regression must be avoided.)

File: ports/wxwidgets/vcpkg-cmake-wrapper.cmake [hunk 2]
```diff
@@ -3,23 +3,27 @@ set(WX_ROOT_DIR "${wxWidgets_ROOT_DIR}")
...
 _find_package(${ARGS})
+
 find_package(ZLIB QUIET)
 find_package(libpng CONFIG QUIET)
 find_package(TIFF QUIET)
 find_package(expat CONFIG QUIET)
 
-if(wxWidgets_LIBRARIES AND NOT wxWidgets_LIBRARIES MATCHES "TIFF::TIFF;png;expat::expat;ZLIB::ZLIB")
+if(wxWidgets_LIBRARIES AND NOT wxWidgets_LIBRARIES MATCHES "^debug;")
+    set(wxWidgets_LIBRARIES "@wxWidgets_LIBRARIES@")
```
[the remark this answers is not in the document]
Comment (jjYBdx4IL) [reply ports/wxwidgets/vcpkg-cmake-wrapper.cmake:19]: I'm not even sure what the intent of the original condition was.

[the remark this answers is not in the document]
Comment (jjYBdx4IL) [reply ports/wxwidgets/vcpkg-cmake-wrapper.cmake:19]: is there a way to add test units to clarify and test use cases?

**Comment by jjYBdx4IL** [comment]
Better?

File: ports/wxwidgets/vcpkg-cmake-wrapper.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjYBdx4IL) [reply ports/wxwidgets/vcpkg-cmake-wrapper.cmake]: Changed. What do you say?

**Comment by jjYBdx4IL** [comment]
> Unfortunately, we lack some CI coverage here, due to WxWidgets' dependency on gtk3 for Linux. There must be manual testing on x64-linux with some port that uses this wrapper. (wxwidgets:x64-linux was fixed only recently, and a immediate regression must be avoided.)

where is the test case?

**Comment by jjYBdx4IL** [comment]
Problem is, the MSVC port currently is totally broken for me. Static build outright fail compilation because debug and release code is mixed. And even if I set dynamic linking in my custom triplet for wxWidgets, it won't pull debug libs and instead link against the release libs which won't be copied into the output directory.

**Comment by dg0yt** [comment]
I wish I could name a test port... It is probably easier to create a minimal testproject yourself.

@talregev Can you test this PR with your Linux project?

**Comment by dg0yt** [comment]
> Problem is, the MSVC port currently is totally broken for me. Static builds outright fail compilation because debug and release code is mixed. And even if I set dynamic linking in my custom triplet for wxWidgets, it won't pull debug libs and instead link against the release libs which also won't be copied into the output directory.

The port has some history of problems and draft PRs. Some outside triggers have been fixed IIRC, and it would be good to resolve the remaining issues. CC @JackBoosY 

Maybe it would be good to (at least temporarily) add a `vcpkg-ci-wxwidgets` test port that tries to build a minimal example app in CI.

I'm willing to test, too. But wxwidgets is not very high on my list, and I use MSVC only in CI.

**Comment by jjYBdx4IL** [comment]
What's the issue with Arzue Pipelines and gtk3? Why not just use github workflows for Linux?

**Comment by jjYBdx4IL** [comment]
> Maybe it would be good to (at least temporarily) add a `vcpkg-ci-wxwidgets` test port that tries to build a minimal example app in CI.

How would that work without gtk3?

How about one (optional) repository on github for each port and which runs against github workflows that allow for more customization?

**Comment by talregev** [comment]
> I wish I could name a test port... It is probably easier to create a minimal testproject yourself.
> 
> @talregev Can you test this PR with your Linux project?
@dg0yt 
I will love to test this PR with the open source project that I am working on,
But currently the state of wxwidgets for my project is that it cannot compile due some of linking problem I mention in previous bug wxwidgets pr.
@JackBoosY promise to fix them and I am still waiting.
Can you solve the problem, then the project can compile with wxwidgets then he can rebase to master, then I can check his pr.

Sounds reasonable?

**Comment by dg0yt** [comment]
> > Maybe it would be good to (at least temporarily) add a `vcpkg-ci-wxwidgets` test port that tries to build a minimal example app in CI.
> 
> How would that work without gtk3?

It wouldn't work now in CI, but a) we could use it locally, and b) there is a gtk3 PR.

> How about one (optional) repository on github for each port and which runs against github workflows that allow for more customization?

When you want to talk about "for each port" or about more GH workflows, you are leaving the scope of this PR. I would avoid it for now.

**Comment by jjYBdx4IL** [comment]
@dg0yt how about simply avoiding the external dependencies and using the builtin ones? Maybe enable the external dependencies only through a feature?

**Comment by jjYBdx4IL** [comment]
> When you want to talk about "for each port" or about more GH workflows, you are leaving the scope of this PR. I would avoid it for now.

Well, we could start with a single, simple repo for wxwidgets test cases. Having a well-defined test suite would make all of this A LOT more efficient. And in future one could simply refuse bug reports without attached test cases....

**Comment by dg0yt** [comment]
@jjYBdx4IL I really wanted to test locally, but with your fork removed, this probably won't happen.

**Comment by talregev** [comment]
@jjYBdx4IL Not sure why you close the PR.
Even you take the code inside your project, it can still import to vcpkg and it will be also for other.

**Comment by jjYBdx4IL** [comment]
> @jjYBdx4IL I really wanted to test locally, but with your fork removed, this probably won't happen.

I pasted the relevant code here. just insert it after your find_package in your CMakeLists. I won't waste weeks on this crap.

**Comment by talregev** [comment]
> > @jjYBdx4IL I really wanted to test locally, but with your fork removed, this probably won't happen.
> 
> I pasted the relevant code here. just insert it after your find_package in your CMakeLists. I won't waste weeks on this crap.

It doesn't work like this, we don't work like this, open source projects don't work like this.

**Comment by jjYBdx4IL** [comment]
> It doesn't work like this, we don't work like this, open source projects doesn't work like this.

What currently doesn't work is your CI system.

**Comment by talregev** [comment]
> > It doesn't work like this, we don't work like this, open source projects doesn't work like this.
> 
> What currently doesn't work is your CI system.

I mean the way of work. We want your contribution, and it slow process. 
So if you can re-open the PR, update it (if needed) then we can check. 
Currently the state of the wxwidgets is not allow me to check it on the ci. with, or without your change.

**Comment by jjYBdx4IL** [comment]
I know what you meant. But this process isn't a "process". It's nonsense.

**Comment by talregev** [comment]
> > It doesn't work like this, we don't work like this, open source projects doesn't work like this.
> 
> What currently doesn't work is your CI system.

This is what is not working on my CI:
https://github.com/microsoft/vcpkg/issues/24038

**Pull request closed without merging**